### PR TITLE
Fire pixel when ad is in viewport

### PR DESCRIPTION
**Pull request by zeantsoi**
:eyeglasses: @ajacksified 

Pursuant to [JIRA #171](https://reddit.atlassian.net/browse/ADS-171):
1. Tracks position of ad unit
2. When center point of ad unit is within the viewport, fire pixel tracker
3. Make impression request by assigning `src` attribute on JS image rather than embedding tag

Scroll tracking calculations and handlers based on code by @danehansen.

### CODE REVIEW

File: views/components/Ad.jsx [hunk 1]
```diff
@@ -51,14 +53,52 @@ class Ad extends React.Component {
...
+    this._onScroll();
+  }
+
+  componentWillUnmount() {
+    this._removeListeners();
+  }
+
+  _removeListeners() {
+    if (this._hasListeners) {
+      window.removeEventListener('scroll', this._onScroll);
```
Comment (ajacksified) [inline views/components/Ad.jsx]: This and resize should be throttled so it isn't calling so many times; try out lodash's `throttle`, maybe at 100ms.

File: views/components/Ad.jsx [hunk 2]
```diff
@@ -51,14 +53,52 @@ class Ad extends React.Component {
...
+    var bottom = top + rect.height;
+    var middle = (top + bottom) / 2;
+    var middleIsAboveBottom = middle < winHeight;
+    var middleIsBelowTop = bottom > 45 + height / 2;
+
+    if(middleIsAboveBottom && middleIsBelowTop) {
+      var srcs=['imp_pixel', 'adserver_imp_pixel'];
+      for (var i = 0, iLen = srcs.length; i < iLen; i++) {
+        var pixel = new Image();
+        pixel.src = this.adObject[srcs[i]];
```
Comment (ajacksified) [inline views/components/Ad.jsx]: This breaks when you don't have an ad (such as on desktop); and why not use `this.state.ad`?

File: views/components/Ad.jsx [hunk 3]
```diff
@@ -51,14 +53,52 @@ class Ad extends React.Component {
   }
 
   componentDidMount () {
-    var ctx = this;
-
     this.getAd().then((ad) => {
+      this.adObject = new models.Link(ad).toJSON();
```
Comment (ajacksified) [inline views/components/Ad.jsx]: Why duplicate this instead of using the same object in state below?

File: views/components/Ad.jsx [hunk 4]
```diff
@@ -51,14 +53,52 @@ class Ad extends React.Component {
...
       return this.setState({
         loaded: true,
-        ad: new models.Link(ad).toJSON(),
+        ad: this.adObject,
       });
     }, () => {});
+
+    window.addEventListener('scroll', this._onScroll);
```
Comment (ajacksified) [inline views/components/Ad.jsx:64]: This and resize should be throttled so it isn't calling so many times; try out lodash's throttle, maybe at 100ms.

**Comment by ajacksified** [comment]
:nail_care: 

Once we get around to lazyloading images, we'll probably move this scroll logic into `app` and create an event. For now, it's fine here, though.

File: views/components/Ad.jsx (hunk 2 above)
Comment (zeantsoi) [inline views/components/Ad.jsx]: I don't believe this breaks when no ads are displayed – I can't see any errors, at any rate.

This component is only invoked by `mutators.jsx` when an ad is available. If there is no ad, the component is never mounted and none of this code should run.

Or am I missing something?

**Comment by zeantsoi** [comment]
:haircut:

**Comment by ajacksified** [comment]
:fish:

**Comment by zeantsoi** [comment]
dde0da2cfe0c90f2adbd65649a61ca3d43f18028

**Pull request closed without merging**